cmd/skopeo/sync.go

```diff
@@ -624,27 +826,61 @@ func (opts *syncOptions) run(args []string, stdout io.Writer) (retErr error) {
 	for _, srcRepo := range srcRepoList {
 		options.SourceCtx = srcRepo.Context
 		for counter, ref := range srcRepo.ImageRefs {
-			var destSuffix string
-			switch ref.Transport() {
-			case docker.Transport:
-				// docker -> dir or docker -> docker
-				destSuffix = ref.DockerReference().String()
-			case directory.Transport:
-				// dir -> docker (we don't allow `dir` -> `dir` sync operations)
-				destSuffix = strings.TrimPrefix(ref.StringWithinTransport(), srcRepo.DirBasePath)
-				if destSuffix == "" {
-					// if source is a full path to an image, have destPath scoped to repo:tag
-					destSuffix = path.Base(srcRepo.DirBasePath)
+			var destRef types.ImageReference
+			if srcRepo.DestinationRef != nil {
+				switch r := ref.DockerReference().(type) {
+				case reference.Digested:
+					destNamedRef, err := reference.ParseNormalizedNamed(srcRepo.DestinationRef.DockerReference().Name())
+					if err != nil {
+						return err
+					}
+
+					destDigestedRef, err := reference.WithDigest(destNamedRef, r.Digest())
+					if err != nil {
+						return err
+					}
+
+					destRef, err = docker.ParseReference(fmt.Sprintf("//%s", destDigestedRef.String()))
+					if err != nil {
+						return err
+					}
+				case reference.NamedTagged:
+					destTaggedRef, err := reference.WithTag(srcRepo.DestinationRef.DockerReference(), r.Tag())
+					if err != nil {
+						return err
+					}
+
+					destRef, err = docker.ParseReference(fmt.Sprintf("//%s", destTaggedRef.String()))
+					if err != nil {
+						return err
+					}
+				default:
+					logrus.Errorf("destination ref type is unsupported for this sync")
+					return fmt.Errorf("destination ref type is unsupported for this sync")
+				}
+			} else {
```

Comment:
Is it really valuable at all to allow specifying a destination on the command line with YAML2? I imagine it’s always going to be the same for the same YAML file, though I’m not quite sure.

(OTOH we don’t support this code path, we need to account for directory destinations in the YAML format.)